sweep/tx_input_set.go

```diff
@@ -542,3 +550,268 @@ func createWalletTxInput(utxo *lnwallet.Utxo) (input.Input, error) {
 		&utxo.OutPoint, witnessType, signDesc, heightHint,
 	), nil
 }
+
+// BudgetInputSet implements the interface `InputSet`. It takes a list of
+// pending inputs which share the same deadline height and groups them into a
+// set conditionally based on their economical values.
+type BudgetInputSet struct {
+	// inputs is the set of inputs that have been added to the set after
+	// considering their economical contribution.
+	inputs []*pendingInput
+
+	// budget is the total budget that can be spent as fees.
+	budget btcutil.Amount
+
+	// deadlineHeight is the height which the inputs in this set must be
+	// confirmed by.
+	deadlineHeight fn.Option[int32]
+}
+
+// Compile-time constraint to ensure budgetInputSet implements InputSet.
+var _ InputSet = (*BudgetInputSet)(nil)
+
+// validateInputs is used when creating new BudgetInputSet to ensure there are
+// no duplicate inputs and they all share the same deadline heights, if set.
+func validateInputs(inputs []pendingInput) error {
+	// Sanity check the input slice to ensure it's non-empty.
+	if len(inputs) == 0 {
+		return fmt.Errorf("inputs slice is empty")
+	}
+
+	// dedupInputs is a map used to track unique outpoints of the inputs.
+	dedupInputs := make(map[*wire.OutPoint]struct{})
```

Comment:
Yeah you can't count on pointer equality to be stable. We should just key off of `wire.OutPoint` directly.

Reply:
> this looks like it will dedupe based off of pointer equality, I wouldn't do this. You should use the raw wire.OutPoint type as the map key.

Will refactor `OutPoint()` to return a concrete type.

> Another way to implement this also would be using fn.Set:

Tried this approach, got an error as `wire.OutPoint` doesn't implement comparable.

Reply:
ok fixed in https://github.com/lightningnetwork/lnd/pull/8581/commits/14d2dfbfb38fc8c230409965171d591a126fbbbc